DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/ifupdown2/test_ifupdown_stp.py

```diff
@@ -0,0 +1,227 @@
+    """
+
+    tgen_dev, dent_devices = await tgen_utils_get_dent_devices_with_tgen(testbed, [], 4)
+    if not tgen_dev or not dent_devices:
+        pytest.skip('The testbed does not have enough dent with tgen connections')
+    dev_name = dent_devices[0].host_name
+    dent_dev = dent_devices[0]
+    tg_ports = tgen_dev.links_dict[dev_name][0]
+    dut_ports = tgen_dev.links_dict[dev_name][1]
+    traffic_duration = 40
+    bridge = 'br0'
+    full_config = ''
+    root_mac = get_rand_mac('00:00:XX:XX:XX:XX')
+    rand_mac = get_rand_mac('00:02:XX:XX:XX:XX')
+    rate_bps = 300_000_000  # 300Mbps
+    traffic = 'L2 traffic'
+    # Forcebly select kernal userspace beforehand
+    rc, out = await dent_dev.run_cmd("sed -i \"/^[#]*MANAGE_MSTPD=/ cMANAGE_MSTPD='n'\" /etc/bridge-stp.conf")
+    assert rc == 0, 'Failed to edit bridge-stp.conf'
+
+    # 1.Prepare ifupdown2 environment config
+    config = {'template_lookuppath': '/etc/network/ifupdown2/templates',
+              'addon_syntax_check': 1,
+              'default_interfaces_configfile': INTERFACES_FILE
+              }
+    rc = await modify_ifupdown_conf(dent_dev, config)
+    assert not rc, 'Failed to prepare ifupdown2 enviroment config'
+
+    # Set the space (user/kernel) according to the STP mode (kernel for STP, userspace for RSTP)
+    if version == 'stp':
+        rc, _ = await dent_dev.run_cmd('if [ ! -z "$(pidof mstpd)" ]; then killall mstpd; fi')
+        assert rc == 0, 'Failed to execute cmd'
+
+    else:
+        rc, _ = await dent_dev.run_cmd('mstpd')
+        rc, out = await dent_dev.run_cmd("sed -i \"/^[#]*MANAGE_MSTPD=/ cMANAGE_MSTPD='y'\" /etc/bridge-stp.conf")
+
+        assert rc == 0, 'Failed to execute cmd'
+
+    # 2.Prepare ifupdown2 config: Add Bridge device, enable stp, add  bridge mac
+    bridge_addr = get_rand_mac('22:XX:XX:XX:XX:XX')
+    full_config += config_bridge_temp(bridge, dut_ports, stp=True, hwaddress=bridge_addr)
+
+    # 3.Verify no errors in ifupdown2 config, apply config, compare running ifupdown2 config vs default config
+    await write_reload_check_ifupdown_config(dent_dev, full_config, config['default_interfaces_configfile'])
+
+    # 4.Verify that bridge device added with 4 member ports and expected bridge mac address
+    # 5.Verify MAC address of the bridge is as set by ifupdown2 configuration
+    out = await IpLink.show(input_data=[{dev_name: [
+        {'device': bridge,
+         'cmd_options': '-j -d'}]}], parse_output=True)
+    assert out[0][dev_name]['rc'] == 0, 'Failed to get port detail'
+    err_msg = f'Address of bridge is not as expected {bridge_addr}'
+    assert out[0][dev_name]['parsed_output'][0]['address'] == bridge_addr, err_msg
+
+    await asyncio.sleep(40)  # Sleep some time for ports to get UP
+    await check_member_devices(dev_name, {bridge: dut_ports})
+
+    # 6. Verify STP is running on the correct space
+    rc, out = await dent_dev.run_cmd(f'cat /sys/class/net/{bridge}/bridge/stp_state')
```

Comment:
Can be checked with ip -j -d link show dev {bridge}, there will stp_state in linkinfo.